### PR TITLE
[Merged by Bors] - feat: tensoring a localization with itself does nothing

### PR DESCRIPTION
**Pull request by alreadydone**
`IsLocalization.tensorSelfAlgEquiv`: If A is a localization of a commutative ring R, then `A ⊗[R] A` is canonically isomorphic to `A` as `A`-algebras via multiplication. To construct the isomorphism we introduce a stronger version of `Algebra.TensorProduct.lmul'`.

Also add some easy theorem (`linearMap_compatibleSMul`) and defs (`distribBaseChange`, `LinearEquiv.baseChange`) and protect `IsLocalization.map_smul`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by erdOne** [comment]
I believe this follows from `B ⊗[A] C = B ⊗[R] C` for localization (or even epi) `R -> A`. Is it hard to get this general statement?

**Comment by alreadydone** [comment]
It should be easy to construct the map from `B ⊗[R] C` to `B ⊗[A] C` using the tensorProduct_compatibleSMul which gives you [TensorProduct.smul_tmul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/TensorProduct/Basic.html#TensorProduct.smul_tmul). (The other map always exists and is surjective given `IsScalarTower R A B/C`; is it somewhere in mathlib?)

File: Mathlib/LinearAlgebra/TensorProduct/Basic.lean [hunk 1]
```diff
@@ -538,6 +538,49 @@ theorem lift_compr₂ (g : P →ₗ[R] Q) : lift (f.compr₂ g) = g.comp (lift f
...
+  map_smul' _ x := x.induction_on (map_zero _) (fun _ _ ↦ by simp [smul_tmul'])
+    fun _ _ h h' ↦ by simpa using congr($h + $h')
+
+theorem mapOfCompatibleSMul_surjective : Function.Surjective (mapOfCompatibleSMul R M N A) :=
+  fun x ↦ x.induction_on (⟨0, map_zero _⟩) (fun m n ↦ ⟨_, mapOfCompatibleSMul_tmul ..⟩)
+    fun _ _ ⟨x, hx⟩ ⟨y, hy⟩ ↦ ⟨x + y, by simpa using congr($hx + $hy)⟩
+
+variable [CompatibleSMul A R M N]
+
+def equivOfCompatibleSMul : M ⊗[A] N ≃ₗ[A] M ⊗[R] N where
```
Comment (alreadydone) [inline Mathlib/LinearAlgebra/TensorProduct/Basic.lean:574]:
```suggestion
def equivOfCompatibleSMul [CompatibleSMul A R M N] : M ⊗[A] N ≃ₗ[A] M ⊗[R] N where
```

File: Mathlib/LinearAlgebra/TensorProduct/Basic.lean [hunk 2]
```diff
@@ -538,6 +538,47 @@ theorem lift_compr₂ (g : P →ₗ[R] Q) : lift (f.compr₂ g) = g.comp (lift f
...
+def mapOfCompatibleSMul' : M ⊗[A] N →ₗ[R] M ⊗[R] N where
+  __ := mapOfCompatibleSMul R M N A
+  map_smul' _ x := x.induction_on (map_zero _) (fun _ _ ↦ by simp [smul_tmul'])
+    fun _ _ h h' ↦ by simpa using congr($h + $h')
+
+theorem mapOfCompatibleSMul_surjective : Function.Surjective (mapOfCompatibleSMul R M N A) :=
+  fun x ↦ x.induction_on (⟨0, map_zero _⟩) (fun m n ↦ ⟨_, mapOfCompatibleSMul_tmul ..⟩)
+    fun _ _ ⟨x, hx⟩ ⟨y, hy⟩ ↦ ⟨x + y, by simpa using congr($hx + $hy)⟩
+
+def equivOfCompatibleSMul [CompatibleSMul A R M N] : M ⊗[A] N ≃ₗ[A] M ⊗[R] N where
```
Comment (alreadydone) [inline Mathlib/LinearAlgebra/TensorProduct/Basic.lean:572]: @erdOne You should be able to get the general isomorphism for epi R -> A by exhibiting the CompatibleSMul instance similar to tensorProduct_compatibleSMul for the localization case. I added the isomorphism for localization as an example.

File: Mathlib/LinearAlgebra/TensorProduct/Basic.lean [hunk 3]
```diff
@@ -1486,3 +1527,5 @@ theorem rTensor_neg (f : N →ₗ[R] P) : (-f).rTensor M = -f.rTensor M := by
 end LinearMap
 
 end Ring
+
+set_option linter.style.longFile 1700
```
Comment (erdOne) [inline Mathlib/LinearAlgebra/TensorProduct/Basic.lean:1531]: What's this?
Should this be here or somewhere in nolint

Comment (alreadydone) [reply Mathlib/LinearAlgebra/TensorProduct/Basic.lean:1531]: It needs to be added (at the end of the file, e.g. [here](https://github.com/leanprover-community/mathlib4/blob/16c9d8c2fba7254dbeb6e41eeb0554c23e5776ec/Mathlib/Data/List/Basic.lean#L2319)) when #lines exceeds 1500

**Comment by erdOne** [comment]
Okay I'm still not sure if it's okay to add these rather than splitting files but I'll leave this up to some maintainer.
Otherwise LGTM
maintainer merge

**Comment by riccardobrasca** [comment]
Can you please fix the error?

File: Mathlib/LinearAlgebra/TensorProduct/Basic.lean (hunk 2 above)
Comment (alreadydone) [inline Mathlib/LinearAlgebra/TensorProduct/Basic.lean:572]:
```suggestion
/-- If the R- and A-actions on M and N satisfy `CompatibleSMul` both ways,
then `M ⊗[A] N` is canonically isomorphic to `M ⊗[R] N`. -/
def equivOfCompatibleSMul [CompatibleSMul A R M N] : M ⊗[A] N ≃ₗ[A] M ⊗[R] N where
```

File: Mathlib/RingTheory/Localization/BaseChange.lean [hunk 4]
```diff
@@ -49,6 +49,39 @@ theorem isLocalizedModule_iff_isBaseChange : IsLocalizedModule S f ↔ IsBaseCha
...
+    simp_rw [algebraMap_smul, smul_tmul', ← smul_assoc, smul_tmul, ← smul_assoc, smul_mk'_self,
+      algebraMap_smul, smul_tmul]
+
+noncomputable example : M₁ ⊗[A] M₂ ≃ₗ[A] M₁ ⊗[R] M₂ :=
+  have := tensorProduct_compatibleSMul S A M₁ M₂
+  equivOfCompatibleSMul R M₁ M₂ A
+
+/-- If A is a localization of a commutative ring R, the tensor product of A with A over R is
+canonically isomorphic as A-algebras to A itself. -/
+noncomputable def tensorSelfAlgEquiv : A ⊗[R] A ≃ₐ[A] A :=
```
Comment (jcommelin) [inline Mathlib/RingTheory/Localization/BaseChange.lean:73]: Do we also have a version of this for modules? E.g. take a Q-vector space, and tensor it over Z with Q, then you get the same vector space.

Comment (alreadydone) [reply Mathlib/RingTheory/Localization/BaseChange.lean:77]: I added this as [an example](https://github.com/leanprover-community/mathlib4/pull/18744/commits/122e2c73e3564888406329e68a5daab6ca22e793#diff-9e7f4ae32a2e11bf6b9c3cd40f4af9f660c7adf15dd698438cc2b54796410b6aR71-R73) and slightly refactored `tensorSelfAlgEquiv`.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Comment by eric-wieser** [comment]
I'm slightly scared to see these `CompatibleSMul` classes which I introduced just to make `Int` actions on modules over semirings easier to work with (`CompatibleSMul \Z R M N`) gain more widespread use; is there a way of thinking of this example as a localization, and does that work with this PR?

(Not an objection to this PR, just a sense of guilt for paving a weird path that you're continuing with)

**Comment by alreadydone** [comment]
As discussed with @erdOne before, what we want is probably a typeclass for epimorphisms of semirings. Localizations, surjective homs and their compositions are examples, but [there are more](https://mathoverflow.net/questions/109/what-do-epimorphisms-of-commutative-rings-look-like). If R → S is an epimorphism and M and N are S-modules (and therefore R-modules), then any R-linear map is automatically an S-linear map, and the S-action on `M ⊗[R] N` can freely switch between the factors, giving the LinearMap/TensorProduct.compatibleSMul instances. Recall R → S being an epimorphism means that to test equality of two homs from S to an arbitrary semiring T, it suffices to test that their composition with R → S are equal. For LinearMap and TensorProduct, the relevant T are `AddMonoid.End (M →+ N)` and `AddMonoid.End (M ⊗[R] N)` respectively (and the two homs are applying `s : S` to `M` or `N`). Now if S = ℤ (which is initial), then there are only one hom from S to an arbitrary T, so there's no need to test and R can be anything (no need to even have a hom from R to S).

If and when we introduce the IsEpi class, I think it would still be easier to provide instances to CompatibleSMul and use existing lemmas for CompatibleSMul like [TensorProduct.smul_tmul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/TensorProduct/Basic.html#TensorProduct.smul_tmul), rather than having our own lemmas for IsEpi.

**Pull request closed without merging**